### PR TITLE
Added max range CLI option for eth_getLogs.

### PR DESCRIPTION
**Pull request by mark-terry**
Signed-off-by: mark-terry <mark.terry@consensys.net>

## PR description
Adds a `--rpc-max-logs-range` CLI option to restrict the maximum block range to query for `eth_getLogs`.

## Fixed Issue(s)

Adds suggestion from #4123 

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 1]
```diff
@@ -1284,6 +1284,11 @@ static class TxPoolOptionGroup {
           "Specifies the static node file containing the static nodes for this node to connect to")
   private final Path staticNodesFile = null;
 
+  @CommandLine.Option(
+      names = {"--rpc-max-logs-range"},
+      description = "Specifies the maximum number of logs to retrieve via RPC")
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1289]: is it a number of logs or a range of blocks?

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 2]
```diff
@@ -1284,6 +1284,11 @@ static class TxPoolOptionGroup {
...
+  private final Long rpcMaxLogsRange = null;
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1290]: should we have a numeric default

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java [hunk 3]
```diff
@@ -651,6 +655,9 @@ public List<LogWithMetadata> matchingLogs(
       final LogsQuery query,
       final Supplier<Boolean> isQueryAlive) {
     try {
+      if (maxLogRange.isPresent() && ((toBlockNumber - fromBlockNumber) > maxLogRange.get())) {
+        throw new Exception("Specified range exceeds maximum range limit");
```
Comment (macfarla) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java:659]: Requested range

File: besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java [hunk 4]
```diff
@@ -1547,6 +1547,20 @@ public void maxpeersOptionMustBeUsed() {
...
+
+    verify(mockRunnerBuilder).rpcMaxLogsRange(longArgumentCaptor.capture());
+    verify(mockRunnerBuilder).build();
+
+    assertThat(longArgumentCaptor.getValue()).isEqualTo(rpcMaxLogsRange);
+
+    assertThat(commandOutput.toString(UTF_8)).isEmpty();
+    assertThat(commandErrorOutput.toString(UTF_8)).isEmpty();
+  }
+
```
Comment (macfarla) [inline besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:1563]: add a test for what happens if you don't specify this option (what's the default behaviour)

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 1 above)
Comment (mark-terry) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java]: Range of blocks. Have updated the description.

File: besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java (hunk 4 above)
Comment (mark-terry) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:1563]: I've removed the empty default and added a numeric one.

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 2 above)
Comment (mark-terry) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java]: Updated.

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 5]
```diff
@@ -1284,6 +1284,12 @@ static class TxPoolOptionGroup {
           "Specifies the static node file containing the static nodes for this node to connect to")
   private final Path staticNodesFile = null;
 
+  @CommandLine.Option(
+      names = {"--rpc-max-logs-range"},
+      description =
+          "Specifies the maximum number of blocks to retrieve logs from via RPC. Value must be > 0")
+  private final Long rpcMaxLogsRange = 0L;
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1291]: would it make it less of a breaking change to have the default be a bigger number?

File: CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,7 @@
 ### Additions and Improvements
 - Explain and improve price validation for London and local transactions during block proposal selection [#4602](https://github.com/hyperledger/besu/pull/4602)
 - Support for ephemeral testnet Shandong.  EIPs are still in flux, besu does not fully sync yet, and the network is subject to restarts. [#//FIXME](https://github.com/hyperledger/besu/pull///FIXME)
+- Added `--rpc-max-logs-range` CLI option to allow limiting the number of blocks queried by `eth_getLogs` RPC API [#4597](https://github.com/hyperledger/besu/pull/4597)
```
Comment (macfarla) [inline CHANGELOG.md:10]: think we need to add a note that this is a breaking change - if you have a workflow that depends on requesting eth_getLogs for a block range that is larger than the new default X

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 7]
```diff
@@ -1284,6 +1284,12 @@ static class TxPoolOptionGroup {
           "Specifies the static node file containing the static nodes for this node to connect to")
   private final Path staticNodesFile = null;
 
+  @CommandLine.Option(
+      names = {"--rpc-max-logs-range"},
```
Comment (macfarla) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1288]: are we only adding this limit for HTTP (not WS?) - if so let's call it rpc-http-max-logs-range - similar to rpc-http-max-active-connections

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 5 above)
Comment (mark-terry) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1291]: 0 = option switched off in this case. It needs to be > 1 to enable the range check.

File: CHANGELOG.md (hunk 6 above)
Comment (mark-terry) [reply CHANGELOG.md:10]: See previous comment - a default hasn't been added _yet_.

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 7 above)
Comment (mark-terry) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1288]: It's not specific to HTTP - it should be effective across HTTP, WS and IPC.

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 5 above)
Comment (macfarla) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1291]: I think we need a non-zero default and doc it as a breaking change (if your workflow relies on more than the default)

**Review by macfarla — APPROVED** [review]
LGTM

**Comment by mark-terry** [comment]
Some data on eth_getLogs output from Mainnet:
1000 block range: ~200Mb output, ~1s processing time
2000 block range: ~400Mb output, ~1.2s processing time
5000 block range: ~1.2Gb output, ~4.5s processing time.

**Review by macfarla — COMMENTED** [review]
@mark-terry you'll need to move the changelog entry to 22.10.2

**Review by Nurbad01 — COMMENTED** [review]
Nur

**Pull request merged**